### PR TITLE
Completed hw4

### PR DESCRIPTION
**Pull request by imOnVacation**

### CODE REVIEW

File: hw4/01-routing.js [hunk 1]
```diff
@@ -1,49 +1,70 @@
...
+app.use((req, res) => {
+    res.status(404).send('404: Page not found')
+    res.end()
+})
 
 app.listen(port, () => {
-  console.log(`Server running at http://localhost:${port}`);
-});
+    console.log(`Server running at http://localhost:${port}`)
+})
```
Comment (sashetov) [inline hw4/01-routing.js:70]: looks good

File: hw4/02-sessions.js [hunk 2]
```diff
@@ -1,17 +1,53 @@
...
+            req.originalUrl
+        }<br><br>Previously visited:<br>${req.session.result.join('<br>')}`
+    )
+})
 
 app.listen(port, () => {
-  console.log(`Server running at http://localhost:${port}`);
-});
+    console.log(`Server running at http://localhost:${port}`)
+})
```
Comment (sashetov) [inline hw4/02-sessions.js:53]: also looks good

File: hw4/03-form.js [hunk 3]
```diff
@@ -1,18 +1,35 @@
...
-});
+app.post('/submit', (req, res) => {
+    // Add your code here
+    const userName = req.body.name
+    const userEmail = req.body.email
+    let userComment = req.body.comment
+    const userLetter = req.body.newsletter
+    let letterOut = ''
+    if (!userComment) userComment = 'n/a'
+    if (!userLetter) letterOut = 'No, thank you'
```
Comment (sashetov) [inline hw4/03-form.js:26]: ok looks correct

File: hw4/04-templates.js [hunk 4]
```diff
@@ -1,64 +1,105 @@
...
+
+    const map = new Map()
+    const countries = []
+
+    axios.get(url).then((response) => {
+        for (let i = 0; i < response.data.length; i += 1) {
+            capital = response.data[i].capital
+            map.set(response.data[i].name.common, capital)
+        }
+        const mapAsc = new Map([...map.entries()].sort())
```
Comment (sashetov) [inline hw4/04-templates.js:40]: nice use of Map

File: hw4/04-templates.js [hunk 5]
```diff
@@ -1,64 +1,105 @@
...
-    results: populous,
-  });
-});
+    // filter the output array for the countries with population of 50 million or more
+    // sort the resulting array to show the results in order of population
+    // map the resulting array into a new array with the country name and formatted population
+
+    const populous = []
+    const map = new Map()
+    const nf = new Intl.NumberFormat('en-US')
```
Comment (sashetov) [inline hw4/04-templates.js:58]: good solution for number format

File: hw4/04-templates.js [hunk 6]
```diff
@@ -1,64 +1,105 @@
...
+    let pop = 0
+    axios.get(url).then((response) => {
+        for (let i = 0; i < response.data.length; i += 1) {
+            pop = response.data[i].population
+            if (parseInt(pop, 10) >= 50000000) {
```
Comment (sashetov) [inline hw4/04-templates.js:63]: i don't think you need to parseInt here as its already a Number but lmk if I'm incorrect

File: hw4/04-templates.js [hunk 7]
```diff
@@ -1,64 +1,105 @@
...
+                map.set(response.data[i].name.common, pop)
+            }
+        }
+        const mapAsc = new Map([...map.entries()].sort((a, b) => b[1] - a[1]))
```
Comment (sashetov) [inline hw4/04-templates.js:67]: nice use of sort and map

File: hw4/04-templates.js [hunk 8]
```diff
@@ -1,64 +1,105 @@
...
+            results: regions,
+        })
+    })
+})
 
 app.listen(port, () => {
-  console.log(`Server running at http://localhost:${port}`);
-});
+    console.log(`Server running at http://localhost:${port}`)
+})
```
Comment (sashetov) [inline hw4/04-templates.js:105]: again, looks good and good use of map

**Review by sashetov — APPROVED** [review]
looks good to merge. I had one small comment regarding parseInt, so if you think I'm correct there maybe remove that but its good to merge otherwise

File: hw4/04-templates.js (hunk 6 above)
Comment (imOnVacation) [reply hw4/04-templates.js]: Hey, you are right! Thanks for catching that, I have removed parseInt.

File: hw4/04-templates.js [hunk 9]
```diff
@@ -1,64 +1,105 @@
...
-  // check for empty data in the output array
-
-  let countries = ['Afghanistan', 'Aland Islands', 'Albania'];
-
-  res.render('page', {
-    heading: 'Countries and Capitals',
-    results: countries,
-  });
-});
+    // map the output array to create an array with country names and capitals
```
Comment (ethanSE) [inline hw4/04-templates.js:29]: I believe that this note means to use Array.prototype.map not Map.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map

To go from an Array of country objects to an array of formatted strings.

**Review by ethanSE — CHANGES_REQUESTED** [review]
I think that there was a confusion here about the note suggesting using map. She is referring to Array.prototype.map unfortunately and not Map. It is confusing that these have a similar name. 

Array.prototype.map
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
Map
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

There's not really any reason to use a Map here and it adds a bit of complexity. If you're not familiar with Array.prototype.map it is definitely worth getting familiar with as it is both very useful and expected knowledge for any work involving javascript.

File: hw4/04-templates.js (hunk 9 above)
Comment (ethanSE) [reply hw4/04-templates.js:29]: I've added this as a change request

**Comment by JeremyIV** [comment]
Everything looks good, nice job!

**Pull request merged**